lxd/devices.go

```diff
@@ -1359,3 +1359,237 @@ func deviceLoadUsb() ([]usbDevice, error) {
+			if bytes.Compare(IBDevResourceBuf, NetDevResourceBuf) != 0 {
+				continue
+			}
+
+			// Now let's find the ports.
+			IBDevID := fmt.Sprintf("/sys/class/net/%s/dev_id", NetDevName)
+			IBDevPort := fmt.Sprintf("/sys/class/net/%s/dev_port", NetDevName)
+			DevIDBuf, err := ioutil.ReadFile(IBDevID)
+			if err != nil {
+				if os.IsNotExist(err) {
+					continue
+				}
+				return nil, err
+			}
+
+			DevIDString := strings.TrimSpace(string(DevIDBuf))
+			DevIDPort, err := strconv.ParseInt(DevIDString, 0, 64)
+			if err != nil {
+				return nil, err
+			}
+
+			DevPort := int64(0)
+			DevPortBuf, err := ioutil.ReadFile(IBDevPort)
+			if err != nil {
+				if !os.IsNotExist(err) {
+					return nil, err
+				}
+			} else {
+				DevPortString := strings.TrimSpace(string(DevPortBuf))
+				DevPort, err = strconv.ParseInt(DevPortString, 0, 64)
+				if err != nil {
+					return nil, err
+				}
+			}
+
+			Port := DevIDPort
+			if DevPort > DevIDPort {
+				Port = DevPort
+			}
+			Port++
+
+			NewIBF := IBF{
+				Port:   Port,
+				Fun:    IBDevName,
+				Device: NetDevName,
+			}
+
+			// identify the /dev/infiniband/uverb<idx> device
+			tmp := []string{}
+			IBUverb := fmt.Sprintf("/sys/class/net/%s/device/infiniband_verbs", NetDevName)
+			fuverb, err := os.Open(IBUverb)
+			if err != nil {
+				if !os.IsNotExist(err) {
+					return nil, os.ErrNotExist
+				}
+			} else {
+				defer fuverb.Close()
+
+				// optional: retrieve all network devices
+				tmp, err = fuverb.Readdirnames(-1)
+				if err != nil {
+					return nil, err
+				}
+
+				if len(tmp) != 1 {
+					return nil, os.ErrNotExist
+				}
+			}
+			for _, v := range tmp {
+				if strings.Index(v, "-") != -1 {
+					return nil, fmt.Errorf("Infiniband character device \"%s\" contains \"-\". Cannot guarantee unique encoding", v)
+				}
+				NewIBF.PerPortDevices = append(NewIBF.PerPortDevices, v)
+			}
+
+			// identify the /dev/infiniband/ucm<idx> device
+			tmp = []string{}
+			IBcm := fmt.Sprintf("/sys/class/net/%s/device/infiniband_ucm", NetDevName)
+			fcm, err := os.Open(IBcm)
+			if err != nil {
+				if !os.IsNotExist(err) {
+					return nil, os.ErrNotExist
+				}
+			} else {
+				defer fcm.Close()
+
+				// optional: retrieve all network devices
+				tmp, err = fcm.Readdirnames(-1)
+				if err != nil {
+					return nil, err
+				}
+
+				if len(tmp) != 1 {
+					return nil, os.ErrNotExist
+				}
+			}
+			for _, v := range tmp {
+				if strings.Index(v, "-") != -1 {
+					return nil, fmt.Errorf("Infiniband character device \"%s\" contains \"-\". Cannot guarantee unique encoding", v)
+				}
+				devPath := fmt.Sprintf("/dev/infiniband/%s", v)
+				NewIBF.PerPortDevices = append(NewIBF.PerPortDevices, devPath)
+			}
+
+			// identify the /dev/infiniband/{issm,umad}<idx> devices
+			IBmad := fmt.Sprintf("/sys/class/net/%s/device/infiniband_mad", NetDevName)
+			ents, err := ioutil.ReadDir(IBmad)
+			if err != nil {
+				if !os.IsNotExist(err) {
+					return nil, err
```

Comment:
Yay, this one looks right, so it's not completely copy/paste through the file :)